tasmota/tasmota_xnrg_energy/xnrg_25_mk_sky_blu.ino

```diff
@@ -0,0 +1,783 @@
+#define MKSB_REG_VOLTAGE_FLOAT    2   // D01 MPPT voltage limit FLOAT, <= restarts charging [mV], SLA battery only
+#define MKSB_REG_OUT_TIMER        3   // D00 Time duration the load gets connected [mh], default: 24h = 24000mh
+#define MKSB_REG_CURRENT          4   //  -  MPPT current limit [mA] e.g. 1000...60000mA 
+                                      //     CAUTION: do not set above hardware-limit 30000/40000/50000/60000
+#define MKSB_REG_BATT_UVP_CUT     5   // D03 Battery UnderVoltageProtection, <=limit cuts load [mV], e.g. 42400mV
+#define MKSB_REG_BATT_UVP_CONN    6   //  ~  Battery UnderVoltageProtection, >=limit reconnects load [mV], typical: D03 + 200mV, e.g. 44400mV
+#define MKSB_REG_COMM_ADDR        7   //  -  Communication Address [mAddress]
+#define MKSB_REG_BATT_TYPE        8   // D04 Battery Type: value 0=SLA, 1=LiPo (2=LiLo, 3=LiFE, 4=LiTo) [mSelect], e.g. 1000=LiPo
+#define MKSB_REG_BATT_CELLS       9   //  -  Battery System: 1...4 * 12V (Read-Only, set at Batt.connection) [m12V], e.g. 4000m12V
+#define MKSB_REG_TOTAL            10  // ^^^ related local parameter at HMI
+
+#define MKSB_COM_RX_GOOD            0
+#define MKSB_COM_ERR_RX_TIMEOUT     1
+#define MKSB_COM_ERR_RX_TOO_SHORT   2
+#define MKSB_COM_ERR_RX_OVERFLOW    3
+#define MKSB_COM_ERR_RX_CRC         4
+#define MKSB_COM_ERR_RX_COMMAND     5
+#define MKSB_COM_ERR_RX_LENGTH      6
+#define MKSB_COM_RX_COUNTER         7
+
+typedef struct mksb_data_t_
+{
+  TasmotaSerial *Serial;
+  char *pRxBuffer;
+  unsigned long time_window;            // for Rx silence detection
+  float temperature;                    // temperature of the charge controller electronics
+#ifdef _WITH_SERIAL_DEBUGGING
+  unsigned long cntTx;                  // requests transmitted to the charge controller
+  unsigned long cntRx[MKSB_COM_RX_COUNTER];
+  uint16_t lastRxCntError;
+#endif
+#ifdef _WITH_CONFIG_REGISTER_SUPPORT
+  uint16_t regs_to_read;                // bit_n = flags register n to read
+  uint16_t regs_to_write;               // bit_n = flags register n to write
+  uint16_t regs_to_report;              // bit_n = flags register n to report (after valid response received)
+  uint16_t regs_value[MKSB_REG_TOTAL];  // temp value storage all known configuration registers
+#endif
+  uint16_t energy_for_import;           // totalizer at charge controller (non-volatile there)
+//  uint8_t mode;                         // status: mode flags
+//  uint8_t error;                        // status: error flags
+  uint16_t status;
+  uint8_t rxCnt;                        // bytecounter at the Rx data buffer
+  uint8_t timeout;                      // active after a request, waiting for reponse
+#ifdef _WITH_ON_OFF_SUPPORT
+  bool actual_state;                    // true = active, false = stop
+  bool target_state;                    // true = active, false = stop
+#endif
+} mksb_data_t;
+
+const char * mksb_register_names[] = {
```

Comment:
Does it work on ESP8266?

If yes, add `PROGMEM` to move the array in Flash instead of RAM